crates/services/executor/src/executor.rs

```diff
@@ -1409,11 +1348,155 @@ where
+                    }) => {
+                        *predicate_gas_used = gas_used;
+                    }
+                    _ => {
+                        debug_assert!(false, "This error is not possible unless VM changes the order of inputs, \
+                        or we added a new predicate inputs.");
+                        return Err(ExecutorError::InvalidTransactionOutcome {
+                            transaction_id: tx_id,
+                        })
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn extra_tx_checks<Tx, T>(
+        &self,
+        mut checked_tx: Checked<Tx>,
+        header: &PartialBlockHeader,
+        tx_st_transaction: &mut StorageTransaction<T>,
+    ) -> ExecutorResult<Checked<Tx>>
+    where
+        Tx: ExecutableTransaction + Send + Sync + 'static,
+        <Tx as IntoChecked>::Metadata: CheckedMetadata,
+        T: KeyValueInspect<Column = Column>,
+    {
+        checked_tx = checked_tx
+            .check_predicates(&CheckPredicateParams::from(&self.consensus_params))
+            .map_err(|e| {
+                ExecutorError::TransactionValidity(TransactionValidityError::Validation(
+                    e,
+                ))
+            })?;
+        debug_assert!(checked_tx.checks().contains(Checks::Predicates));
+
+        self.verify_inputs_exist_and_values_match(
+            tx_st_transaction,
+            checked_tx.transaction().inputs(),
+            header.da_height,
+        )?;
+        checked_tx = checked_tx
+            .check_signatures(&self.consensus_params.chain_id())
+            .map_err(TransactionValidityError::from)?;
+        debug_assert!(checked_tx.checks().contains(Checks::Signatures));
+        Ok(checked_tx)
+    }
+
+    fn attempt_tx_execution_with_vm<Tx, T>(
+        &self,
+        checked_tx: &Checked<Tx>,
+        header: &PartialBlockHeader,
+        coinbase_contract_id: ContractId,
+        gas_price: Word,
+        tx_st_transaction: &mut StorageTransaction<T>,
+    ) -> ExecutorResult<(bool, ProgramState, Tx, Vec<Receipt>)>
+    where
+        Tx: ExecutableTransaction + Cacheable,
+        <Tx as IntoChecked>::Metadata: CheckedMetadata,
+        T: KeyValueInspect<Column = Column>,
+    {
+        let tx_id = checked_tx.id();
+
+        let mut sub_block_db_commit = tx_st_transaction
+            .read_transaction()
+            .with_policy(ConflictPolicy::Overwrite);
+
+        let vm_db = VmStorage::new(
+            &mut sub_block_db_commit,
+            &header.consensus,
+            &header.application,
+            coinbase_contract_id,
+        );
+
+        let mut vm = Interpreter::with_storage(
+            vm_db,
+            InterpreterParams::new(gas_price, &self.consensus_params),
+        );
+
+        let gas_costs = self.consensus_params.gas_costs();
+        let fee_params = self.consensus_params.fee_params();
+
+        let ready_tx = checked_tx
+            .clone()
```

Comment:
Now(because we don't need to compare `checked_tx` with `tx`) you can create a vector of `predicate_gas_used` from inputs and then you don't need to `checked_tx.clone()`=)